core/src/builder/parachain.rs

```diff
@@ -185,103 +511,298 @@ where
-	pub fn build_block(&mut self) -> Result<(), ()> {
-		assert!(self.next.is_none());
-
-		let provider = self
-			.with_state(|| {
+	pub fn build_block(&mut self) -> Result<(), Error<Block>> {
+		let provider =
+			self.with_state(|| {
 				futures::executor::block_on(self.cidp.create_inherent_data_providers(
 					self.builder.latest_block(),
 					ExtraArgs::extra(),
 				))
-				.unwrap()
-			})
-			.unwrap();
-
-		let parent = self.builder.latest_header();
-		let inherents = futures::executor::block_on(provider.create_inherent_data()).unwrap();
-		let digest = self
-			.with_state(|| {
-				futures::executor::block_on(self.dp.create_digest(parent, inherents.clone()))
-					.unwrap()
-			})
-			.unwrap();
+				.map_err(|e| {
+					tracing::error!(
+						target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+						error = ?e,
+						"Could not create inherent data providers."
+					);
+
+					Error::InherentDataProvidersCreation(e)
+				})
+			})??;
+
+		let inherents =
+			futures::executor::block_on(provider.create_inherent_data()).map_err(|e| {
+				tracing::error!(
+					target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+					error = ?e,
+					"Could not create inherent data."
+				);
+
+				Error::InherentDataCreation(e.into())
+			})?;
+
+		let parent = self.builder.latest_header().map_err(|e| {
+			tracing::error!(
+				target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+				error = ?e,
+				"Could not retrieve latest header."
+			);
+
+			Error::CoreBuilder(e.into())
+		})?;
+
+		let digest = self.with_state(|| {
+			futures::executor::block_on(self.dp.create_digest(parent, inherents.clone())).map_err(
+				|e| {
+					tracing::error!(
+						target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+						error = ?e,
+						"Could not create digest."
+					);
+
+					Error::DigestCreation(e.into())
+				},
+			)
+		})??;
+
+		let Proposal { block, proof, .. } = self
+			.builder
+			.build_block(
+				self.builder.handle(),
+				inherents,
+				digest,
+				Duration::from_secs(60),
+				6_000_000,
+			)
+			.map_err(|e| {
+				tracing::error!(
+					target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+					error = ?e,
+					"Could not create inherent data."
+				);
+
+				Error::CoreBuilder(e.into())
+			})?;
+
+		// As collation info needs latest state in db we import without finalizing here already.
+		let (header, body) = block.clone().deconstruct();
+		let mut params = BlockImportParams::new(BlockOrigin::Own, header);
+		params.body = Some(body);
+		params.fork_choice = Some(ForkChoiceStrategy::Custom(false));
+
+		self.import_block_with_params(params)?;
+
+		// self.set_next_block_and_import(Some((block, proof)))?;
+		let binding = self.next_block.clone();
+		let mut next_block = binding.lock().map_err(|e| {
+			tracing::error!(
+				target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+				error = ?e,
+				"Lock for next block is poisoned.",
+			);
+
+			Error::NextBlockLockPoisoned(InnerError::from(e.to_string()))
+		})?;
+
+		*next_block = Some((block, proof));
```

Comment:
We might need to consider racing conditions between the collator and the builder if we make the whole stuff async. But for now it should work fine.

Reply:
The collator also holds a `Arc<Mutex<Option<(Block, StorageProof)>>>`, I think we are safe in that regard,

Reply:
From a computational point that is true. I am just not sure, if the internal logic has problems with that.